nixos/modules/services/web-apps/keycloak.nix

```diff
@@ -0,0 +1,441 @@
+                echo "Database password file ${cfg.databasePasswordFile} not found! Aborting." >&2
+                exit 1
+            fi
+
+            PSQL=${config.services.postgresql.package}/bin/psql
+
+            db_password="$(<'${cfg.databasePasswordFile}')"
+
+            $PSQL -tAc "SELECT 1 FROM pg_roles WHERE rolname='${cfg.databaseUsername}'" | grep -q 1 || $PSQL -tAc "CREATE ROLE ${cfg.databaseUsername} WITH LOGIN PASSWORD '$db_password' CREATEDB"
+
+            $PSQL -tAc "SELECT 1 FROM pg_database WHERE datname = 'keycloak'" | grep -q 1 || $PSQL -tAc 'CREATE DATABASE "keycloak" OWNER "${cfg.databaseUsername}"'
+            current_owner=$($PSQL -tAc "SELECT pg_catalog.pg_get_userbyid(datdba) FROM pg_catalog.pg_database WHERE datname = 'keycloak'")
+            if [[ "$current_owner" != "${cfg.databaseUsername}" ]]; then
+                $PSQL -tAc 'ALTER DATABASE "keycloak" OWNER TO "${cfg.databaseUsername}"'
+                if [[ -e "${config.services.postgresql.dataDir}/.reassigning_keycloak" ]]; then
+                    echo "Reassigning ownership of database keycloak to user ${cfg.databaseUsername} failed on last boot. Failing..." >&2
+                    exit 1
+                fi
+                touch "${config.services.postgresql.dataDir}/.reassigning_keycloak"
+                $PSQL "keycloak" -tAc "REASSIGN OWNED BY \"$current_owner\" TO \"${cfg.databaseUsername}\""
+                rm "${config.services.postgresql.dataDir}/.reassigning_keycloak"
+            fi
+          '';
+        };
+
+        systemd.services.keycloak = {
+          after = lib.optionals databaseActuallyCreateLocally [
+            "keycloakDatabaseInit.service" "postgresql.service"
+          ];
+          bindsTo = lib.optionals databaseActuallyCreateLocally [
+            "keycloakDatabaseInit.service" "postgresql.service"
+          ];
+          wantedBy = [ "multi-user.target" ];
+          environment = {
+            JBOSS_LOG_DIR = "/var/log/keycloak";
+            JBOSS_BASE_DIR = "/var/lib/keycloak";
+            JBOSS_MODULEPATH = "${cfg.package}/modules";
+          };
+          serviceConfig = {
+            ExecStartPre = systemdPrefixScript "startPre" "+" (''
+              set -eu
+
+              if ! [[ -d /var/lib/keycloak/configuration ]]; then
+                  cp -r ${cfg.package}/standalone/configuration /var/lib/keycloak/
+              fi
+
+              if ! [[ -e ${cfg.databasePasswordFile} ]]; then
+                  echo "Database password file ${cfg.databasePasswordFile} not found! Aborting." >&2
+                  exit 1
+              fi
+              cp ${keycloakConfig} /var/lib/keycloak/configuration/standalone.xml
+              db_password="$(<'${cfg.databasePasswordFile}')"
+              ${pkgs.replace}/bin/replace-literal -fe '@db-password@' "$db_password" /var/lib/keycloak/configuration/standalone.xml
+
+              mkdir -p /var/lib/keycloak/{deployments,ssl}
```

Comment:
I would prefer declarative [StateDirectory](https://www.freedesktop.org/software/systemd/man/systemd.exec.html) over imperative `bash` statements.

Reply:
I rewrote this; it now uses `RuntimeDirectory` for all required directories.